cf-testd/cf-testd.c

```diff
@@ -607,16 +620,153 @@ int main(int argc, char *argv[])
      * cause a segfault or something */
     ServerTLSDeInitialize(&(config->priv_key), &(config->pub_key), &(config->ssl_ctx));
 
-    return r;
+    return NULL;
+}
+
+static void HandleSignal(int signum)
+{
+    switch (signum)
+    {
+    case SIGTERM:
+    case SIGINT:
+        // flush all logging before process ends.
+        fflush(stdout);
+        fprintf(stderr, "Terminating...\n");
+        TERMINATE = true;
+        break;
+    default:
+        break;
+    }
+}
+
+/**
+ * @param ip_str string representation of an IPv4 address (the usual one, with
+ *               4 numbers separated by dots)
+ */
+static char *IncrementIPaddress(const char *ip_str)
+{
+    struct in_addr ip;
+    int ret = inet_aton(ip_str, &ip);
+    if (ret == 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to parse address: '%s'", ip_str);
+        return NULL;
+    }
+
+    int step = 1;
+    char *last_dot = strrchr(ip_str, '.');
+    if (strcmp(last_dot + 1, "255") == 0)
+    {
+        /* avoid the network address (ending with 0) */
+        step = 2;
+    }
+    else if (strcmp(last_dot + 1, "254") == 0)
+    {
+        /* avoid the broadcast address and the network address */
+        step = 3;
+    }
+
+    uint32_t ip_num = ntohl(ip.s_addr);
+    ip_num += step;
+    ip.s_addr = htonl(ip_num);
+
+    return xstrdup(inet_ntoa(ip));
+}
+
+int main(int argc, char *argv[])
+{
+    signal(SIGINT, HandleSignal);
+    signal(SIGTERM, HandleSignal);
+
+    Log(LOG_LEVEL_VERBOSE, "Starting cf-testd");
+    CryptoInitialize();         /* TODO: remove */
+    cfnet_init(NULL, NULL);
+    MakeSignalPipe();
+
+    long n_threads = 1;
+    CFTestD_Config *config = CFTestD_CheckOpts(argc, argv, &n_threads);
+    if (config->address == NULL)
+    {
+        /* default to localhost */
+        config->address = xstrdup("127.0.0.1");
+    }
+
+    CFTestD_Config **thread_configs = (CFTestD_Config**) xcalloc(n_threads, sizeof(CFTestD_Config*));
+    for (int i = 0; i < n_threads; i++)
+    {
+        thread_configs[i] = (CFTestD_Config*) xmalloc(sizeof(CFTestD_Config));
+
+        if (config->report_file != NULL && strstr(config->report_file, "%d") != NULL)
+        {
+            asprintf(&(thread_configs[i]->report_file), config->report_file, i);
+        }
+        else
+        {
+            thread_configs[i]->report_file = SafeStringDuplicate(config->report_file);
+        }
+
+        if (config->key_file != NULL && strstr(config->key_file, "%d") != NULL)
+        {
+            asprintf(&(thread_configs[i]->key_file), config->key_file, i);
+        }
+        else
+        {
+            thread_configs[i]->key_file = SafeStringDuplicate(config->key_file);
+        }
+
+        if (i == 0)
+        {
+            thread_configs[i]->address = xstrdup(config->address);
+        }
+        else
+        {
+            thread_configs[i]->address = IncrementIPaddress(thread_configs[i-1]->address);
+        }
+    }
+
+    /* CFTestD_ConfigDestroy(config); */
+
```

Comment:
go ahead and delete this comment right?

Reply:
Actually, it should only be uncommented. :)